### PR TITLE
Main

### PR DESCRIPTION
**Pull request by OlegMR0**

### CODE REVIEW

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 1]
```diff
@@ -5,10 +5,39 @@
...
+                && checkUserLoginForLength(user)
+                && checkUserPasswordForLength(user);
+        if (correctUser) {
+            storageDao.add(user);
+            return user;
+        }
+        throw new InvalidUserException();
+    }
+
+    boolean checkUserLoginForIdentity(User user) {
```
Comment (kozhukhovsky) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:28]:
```suggestion
    private boolean checkUserLoginForIdentity(User user) {
```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    @BeforeEach
+    void setUp() {
+        defaultCorrectUser = new User();
+        defaultCorrectUser.setAge(20);
+        defaultCorrectUser.setLogin("correct login");
+        defaultCorrectUser.setPassword("correctPassword");
+    }
+
+    @Test
+    void register_NullUser_NotOk() {
```
Comment (kozhukhovsky) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:25]:
```suggestion
    void register_nullUser_notOk() {
```

This style must be observed in all methods.

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        User user = null;
+        assertThrows(InvalidUserException.class,
+                () -> REGISTRATION_SERVICE.register(user));
```
Comment (kozhukhovsky) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:28]: image

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+import core.basesyntax.db.Storage;
+import core.basesyntax.model.User;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+class RegistrationServiceImplTest {
```
Comment (kozhukhovsky) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:11]: image

**Review by kozhukhovsky — CHANGES_REQUESTED** [review]

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java (hunk 3 above)
Comment (OlegMR0) [reply src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:29]: I can't understand properly. If I replaced with a lower method, exception will not throw.

File: src/main/java/core/basesyntax/service/InvalidUserException.java [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package core.basesyntax.service;
+
+public class InvalidUserException extends RuntimeException {
+    public InvalidUserException() {
+        super("This user can't be register.");
```
Comment (okuzan) [inline src/main/java/core/basesyntax/service/InvalidUserException.java:5]:
```suggestion
        super("This user can't be registered.");

```

Comment (okuzan) [inline src/main/java/core/basesyntax/service/InvalidUserException.java:5]:
```suggestion
    public InvalidUserException(String message) {
        super(message);

```

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 6]
```diff
@@ -5,10 +5,52 @@
...
+    private static final int MINIMAL_LENGTH = 6;
+    private static final int MINIMAL_AGE = 18;
     private final StorageDao storageDao = new StorageDaoImpl();
 
     @Override
-    public User register(User user) {
-        return null;
+    public User register(User user) throws RuntimeException {
+        if (user == null) {
+            throw new InvalidUserException();
```
Comment (okuzan) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:15]: would be nice to have messages which would make debugging easier by specifying what exactly is wrong. For example here, "user can't be null"

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java (hunk 3 above)
Comment (okuzan) [reply src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:29]: This checklist bullet point pertains to test for registering already existing user

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+
+    @Test
+    void register_existingUser_NotOk() {
+        Storage.people.add(defaultCorrectUser);
+        User user = new User();
+        user.setLogin(defaultCorrectUser.getLogin());
+        user.setAge(20);
+        user.setPassword("abcdefg");
+        assertThrows(InvalidUserException.class,
+                () -> REGISTRATION_SERVICE.register(defaultCorrectUser));
```
Comment (okuzan) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:139]:
```suggestion
        assertThrows(InvalidUserException.class,
                () -> REGISTRATION_SERVICE.register(defaultCorrectUser));

```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+import core.basesyntax.db.Storage;
+import core.basesyntax.model.User;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+class RegistrationServiceImplTest {
+    private static final RegistrationServiceImpl REGISTRATION_SERVICE
+            = new RegistrationServiceImpl();
```
Comment (okuzan) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:13]: more idiomatic approach would be making it non static field, initialized in @BeforeEach

**Review by okuzan — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 9]
```diff
@@ -5,10 +5,62 @@
...
+    private static final int MINIMAL_AGE = 18;
     private final StorageDao storageDao = new StorageDaoImpl();
 
     @Override
-    public User register(User user) {
-        return null;
+    public User register(User user) throws RuntimeException {
+        if (user == null) {
+            throw new InvalidUserException("User can't be null!");
+        }
```
Comment (uliana-andrushchuk) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:16]: add checks `user.getAge() == null` `user.getPassword() == null`

**Review by uliana-andrushchuk — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java (hunk 9 above)
Comment (OlegMR0) [reply src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:16]: There are already boolean checks below. Or what do you mean?

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 10]
```diff
@@ -5,10 +5,62 @@
...
+            return false;
+        }
+        return user.getPassword().length() >= MINIMAL_LENGTH;
+    }
+
+    private boolean checkUserForAge(User user) {
+        if (user.getAge() == null) {
+            return false;
+        }
+        return user.getAge() >= MINIMAL_AGE;
```
Comment (okuzan) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:64]: combine conditions, you don't need if blocks in checker methods

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 11]
```diff
@@ -5,10 +5,62 @@
...
+            exception.append("Incorrect login!").append(System.lineSeparator());
+        }
+        if (!checkUserPasswordForLength(user)) {
+            exception.append("Incorrect password!").append(System.lineSeparator());
+        }
+        if (exception.length() == 0) {
+            storageDao.add(user);
+            return user;
+        }
+        throw new InvalidUserException(exception.toString());
```
Comment (okuzan) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:35]: well if you opt to do it this way, then to make it better: Create a list of string     
List<String> errorMessages = new ArrayList<>();
then add messages one by one
then         throw new InvalidUserException(String.join(System.lineSeparator(), errorMessages));

**Review by okuzan — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java (hunk 11 above)
Comment (OlegMR0) [reply src/main/java/core/basesyntax/service/RegistrationServiceImpl.java]: Can you explain please why this method is better?

**Review by kozhukhovsky — APPROVED** [review]

**Pull request closed without merging**